### PR TITLE
feat: add support for ManiaPlanet4 (MP4) aka TrackMania 2.

### PR DESCRIPTION
**Pull request by Atheuz**
Implemented by adding `VehicleMP4.as` that contains an interface to get the data we want.

Also required adjustments to `Dashboard.as` because `CSceneVehicleVis` can't be used in MP4, as such it was replaced with `CSceneVehicleVisInner`.
We needed to do this because because `CSceneVehicleVis` is a class in the game and doesn't have the data we need.

Additionally, we use the currently viewed player's script API to get the data we want.

Note that a lot of information present in TM2020 or TM Ultra is unavailable in MP4 (or at least I don't know where it is):
- Wheel rotation, rotation speed, steer angle, slip coefficient
- World velocity
- Side speed

Note that spectated players don't work fully either, specifically the only information available for a spectated player is their Speed and Engine RPM.

### CODE REVIEW

**Comment by codecat** [comment]
Nice! Does this work when playing in a server too? Why does it not work with spectating? Can we make it work somehow?

**Comment by Atheuz** [comment]
@codecat 
>  Does this work when playing in a server too?
Yes.

> Why does it not work with spectating?
Because this: `playground.GameTerminals[0].GUIPlayer;` doesn't return the spectated player, it returns the spectating player. 
`playground.Players` contains all players on a server, but we'd have to determine what player is being spectated. However I noticed that when I'm looking at it it doesn't really seem like a lot of the values are correct: InputGasPedal or InputSteer or InputIsBreaking is always 0 or false. Also it seems like other values are not updated very frequently.

**Comment by codecat** [comment]
Still better than nothing I guess. See also this comment for info about spectating players: https://github.com/codecat/tm-dashboard/issues/2#issuecomment-874227402

**Comment by Atheuz** [comment]
Changed to using CurrentPlayground.Interface.UiPlayer0 for data instead, because that does allow you to get info about spectated players, but a lot is missing to the point where the only things that are actually available are Engine RPM and Speed.

**Comment by Atheuz** [comment]
@codecat I can try to access the memory that you outlined in the issue (https://github.com/codecat/tm-dashboard/issues/2), but this is touching things I'm not really competent with, I'm good at modifying plugins that just access the Nods that are documented in https://mp4.openplanet.nl/, not so much when it comes to futzing around with memory and offsets.

File: Source/Dashboard.as [hunk 1]
```diff
@@ -36,6 +36,15 @@ class Dashboard
...
 		return playground.LocalPlayerMobil;
 	}
+#elif MP4
+	CGamePlayer@ GetViewingPlayer()
+	{
+		auto playground = GetApp().CurrentPlayground;
+		if (playground is null || playground.Interface is null) {
+			return null;
+		}
+        return cast<CTrackManiaRaceInterface>(playground.Interface).UiPlayer0;
```
Comment (codecat) [inline Source/Dashboard.as:46]: Inconsistent indentation

**Review by codecat — CHANGES_REQUESTED** [review]
Minor style nitpick

**Comment by codecat** [comment]
> I can try to access the memory that you outlined in the issue (#2)

Even with the `Dev` API you likely won't be able to get much info from it, as described in the other comment.

> the only things that are actually available are Engine RPM and Speed

If that's the case, then it should probably just not be available for spectating. That's unfortunate, but there's not really anything we can do about that. I'd say revert that last change to remove support for spectating again, and just leave it to the local player.

What's missing on the local player values right now? Is it like in Turbo?

**Comment by Atheuz** [comment]
> 
> 
> > I can try to access the memory that you outlined in the issue (#2)
> 
> Even with the `Dev` API you likely won't be able to get much info from it, as described in the other comment.
> 
> > the only things that are actually available are Engine RPM and Speed
> 
> If that's the case, then it should probably just not be available for spectating. That's unfortunate, but there's not really anything we can do about that. I'd say revert that last change to remove support for spectating again, and just leave it to the local player.
> 
> What's missing on the local player values right now? Is it like in Turbo?

So as I understand it, we have these values in Turbo, I've marked the ones that are available through the Script API that I'm accessing for MP4.

- [X] float get_InputSteer()
- [X] float get_InputGasPedal()
- [X] float get_InputBrakePedal()
- [X] bool  get_InputIsBraking()
- [X] uint  get_CurGear()
- [X] float get_RPM()
- [X] float get_FrontSpeed()
- [ ] float get_SideSpeed()
- [X] vec3  get_Position()
- [ ] vec3  get_WorldVel()
- [ ] float get_FLWheelRot()
- [ ] float get_FLWheelRotSpeed()
- [ ] float get_FLSteerAngle()
- [ ] float get_FLSlipCoef()
- [ ] float get_FRWheelRot()
- [ ] float get_FRWheelRotSpeed()
- [ ] float get_FRSteerAngle()
- [ ] float get_FRSlipCoef()
- [ ] float get_RLWheelRot()
- [ ] float get_RLWheelRotSpeed()
- [ ] float get_RLSteerAngle()
- [ ] float get_RLSlipCoef()
- [ ] float get_RRWheelRot()
- [ ] float get_RRWheelRotSpeed()
- [ ] float get_RRSteerAngle()
- [ ] float get_RRSlipCoef()

In summary, missing: SideSpeed, WorldVel, Wheel info

**Review by codecat — APPROVED** [review]

**Comment by codecat** [comment]
Thanks! I'll test this later tonight!

**Comment by codecat** [comment]
Thank you!! ❤️
